### ISSUE TITLE
Unable to format event with trailing text after it has been flattened

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8003]
|<img alt="pauls's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| pauls reported|
|-|-|
|Trac ID|trac#8003|
|Type|defect|
|Created|2015-08-20 13:33:03Z|
|Branch|https://github.com/twisted/twisted/tree/log-flatten-trailing-8003|

After a log event that has trailing text has been flattened by flattenEvent, flatFormat will try to look up a field value of None. Therefore, raising a KeyError and causing the event to not be formatted.

One case where this shows up is when formatting events that were serialized using the JSON log observer.

The issue is caused by not handling the case where the field name returned by string.Formatter.parse() is None.

I have attached a test script and a patch.

Attachments:

* [flatformat.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b5a/b5ac90fbdd1355438a65edd2fabe14e9fcca10ea/578cd29dc52ef5ca3fcc2468c58c9575efd50ee4.patch) (686 bytes) - added by pauls on 2015-08-20 13:33:52Z - 
* [test.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b5a/b5ac90fbdd1355438a65edd2fabe14e9fcca10ea/a36e355ddb203d7d4133221f339dc406cb9f480f.py) (453 bytes) - added by pauls on 2015-08-20 13:34:23Z - 
* [flatten-trailing-text-8003-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b5a/b5ac90fbdd1355438a65edd2fabe14e9fcca10ea/f8d239bb504fcef3c5afd7cfe1b459a73c463a4a.patch) (1513 bytes) - added by pauls on 2015-08-26 15:52:00Z - 
* [flatten-trailing-text-8003-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b5a/b5ac90fbdd1355438a65edd2fabe14e9fcca10ea/4bd3aa034c27fcf5e1951314185a798e790d2e1e.patch) (2048 bytes) - added by pauls on 2015-08-27 14:15:51Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8003 8003
type__defect defect
reporter__pauls pauls
priority__normal normal
milestone__ 
branch__branches_log_flatten_trailing_8003 branches/log-flatten-trailing-8003
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1440077583452359 1440077583452359
changetime__1440694060889367 1440694060889367
version__None None
owner__pauls pauls

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="pauls's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|pauls removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban set owner to pauls|
|-|-|

Hi Paul and many thanks for the patch and sorry for the delay.

Please note that in order for your patch to be accepted, it need to be submitted as a single patch which includes automated tests.

With an automated test we would make sure that the feature that you need / or the bug will stay fixed forever, even if we refactor the code.

Let me know if you need help with writing the automated tests.

Here is the brief docs about submitting a patch:
http://twistedmatrix.com/trac/wiki/TwistedDevelopment#SubmittingaPatch

Here you can find some documentation about Twisted's dev policy: http://twistedmatrix.com/documents/current/core/development/policy/index.html

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="pauls's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|pauls removed owner|
|-|-|

Uploaded new patch that includes unit test (flatten-trailing-text-8003-2.patch​)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

(In [45560]) Branching to log-flatten-trailing-8003.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to pauls|
|-|-|

I have applied your patch and sent it to the test builders.

Please note that your patch needs a topfile file with the content to be included in the release notes.

I have committed a placeholder topfile. Please use it to create a final version.

here you can find more info about the topfiles - http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

Your test.pt imports from a private module _flatten so I don't know how end users are affected by this bug.

End users are not supposed to import from private module and the release notes should only contain information about the public methods.

-------

The docstring of your test talks about 'L{flatFormat}' but I see that it uses flattenEvent.

--------

Please update the topfile to talk about what is fixed from the end user perspective, and update the test docstring and resubmit.

Please also check the results of bulildbot... I hope that no test will fail

Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

Here are some formating tests which failed

```
************* Module twisted.logger._flatten
C0301:175,0: Line too long (80/79)
************* Module twisted.logger.test.test_flatten
W9012:223,0: Expected 2 blank lines, found 1
```

**Comment by twisted-trac** [comment]
|<img alt="pauls's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|pauls removed owner|
|-|-|

I uploaded a new patch. I based the patch on branches/log-flatten-trailing-8003. This is my first contribution to the project so I was not sure if the patch should be based on trunk or the branch you created.

Regarding test.py: A user would run into this issue if they used twisted.logger.eventAsJSON (it calls flattenEvent) then used twisted.logger.formatEvent on the de-serialized result (eventFromJSON) or the original event. I used the private module just to simplify my test script. I can update it to use eventAsJSON if needed.

Thanks,
Paul

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to pauls|
|-|-|

A patch on the attached branch is great :) Many thanks.

No need to update the test.py file... we now have an automated test with a valid docstring and we also have a news file explaining what was the problem from the user perspective :)

----------

I am sending the patch to buildbot and if all is green (and it should be) I will merge.

I have changed the name from test_formatFlatEventTrailingText to test_formatEventFlatTrailingText so that it is easier to see that it is about formatEvent ... trials does not support running tests based on a patter ... but maybe in the future it will. With this coding convention you can run `trial -e '.*formatEvent.*'` and it will run all tests related to formatEvent.

I have also broke the tests into Arrange Act Assert pattern http://c2.com/cgi/wiki?ArrangeActAssert  as I find them easier to read.

Thanks for your contribution!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set status to `closed`|
|-|-|

(In [45568]) Merge log-flatten-trailing-8003: Fix twisted.logger.formatEvent for flattened events with trailing text.

Author: pauls
Reviewer: adiroiban
Fixes: [#8003](https://github.com/twisted/twisted/issues/8003)
